File: testbed/test/cruise/statemachine/test/MixinTest.java
package cruise.statemachine.test;

import org.junit.*;

public class MixinTest
{

  @Test
  public void SharedEvents()
  {
    CourseM course = new CourseM();
    course.flip();
    
    Assert.assertEquals(CourseM.One.Off,course.getOne());
    Assert.assertEquals(CourseM.Two.Fast, course.getTwo());
  }

  @Test
  public void AddEvents()
  {
    CourseN course = new CourseN();
    course.flip();
    Assert.assertEquals(CourseN.One.Off,course.getOne());
  }
  
  @Test
  public void RemoveEvent()
  {
    CourseO course = new CourseO();
    try
    {
      Assert.assertEquals(null, course.getClass().getMethod("flip"));
      Assert.fail("flip should not exist");
    }
    catch (NoSuchMethodException e)
    {
    }
  }
  
  @Test
  public void AddState()
  {
    Assert.assertEquals(CourseP.One.Amber,CourseP.One.valueOf("Amber"));
    CourseP course = new CourseP();
    course.flip();
    Assert.assertEquals(CourseP.One.Amber,course.getOne());
  }
  
  @Test(expected = IllegalArgumentException.class)  
  public void RemoveState()
  {
    CourseP.One.valueOf("Off");
  }

  
}

File: testbed_php/test/statemachine/MixinTest.php
<?php

class MixinTest extends UnitTestCase
{

  public function test_SharedEvents()
  {
    $course = new CourseM();
    $course->flip();
    
    $this->assertEqual("OneOff",$course->getOne());
    $this->assertEqual("TwoFast", $course->getTwo());
  }

  public function test_AddEvents()
  {
    $course = new CourseN();
    $course->flip();
    $this->assertEqual("OneOff",$course->getOne());
  }
  
  public function test_RemoveEvent()
  {
    $course = new CourseO();
	  $this->assertEqual(false, method_exists($course,"flip"));
  }
  
  public function test_AddState()
  {
    $course = new CourseP();
    $course->flip();
	  $this->assertEqual("OneAmber",$course->getOne());
  }
  
  public function test_RemoveState()
  {
    $course = new ReflectionClass('CourseP');
  	try
  	{
  	  $course->getStaticPropertyValue('$OneOff');
      $this->fail("OneOff should be removed");
  	} catch (Exception $e) {
    }
	}
}

